iroha/src/lib.rs

```diff
@@ -16,36 +16,96 @@ pub mod tx;
 pub mod wsv;
 
 use crate::{
-    block::Blockchain, config::Configuration, kura::Kura, sumeragi::Sumeragi, torii::Torii,
-    wsv::World,
+    block::Blockchain, config::Configuration, kura::Kura, prelude::*, queue::Queue,
+    sumeragi::Sumeragi, torii::Torii,
+};
+use futures::{
+    channel::mpsc::{self, UnboundedReceiver, UnboundedSender},
+    executor::ThreadPool,
+    lock::Mutex,
 };
-use futures::channel::mpsc;
 use parity_scale_codec::{Decode, Encode};
-use std::path::Path;
+use std::{path::Path, sync::Arc, time::Instant};
+
+pub type BlockSender = UnboundedSender<Block>;
+pub type TransactionSender = UnboundedSender<Transaction>;
+pub type TransactionReceiver = UnboundedReceiver<Transaction>;
+pub type BlockReceiver = UnboundedReceiver<Block>;
 
 pub struct Iroha {
-    torii: Torii,
+    torii: Arc<Mutex<Torii>>,
+    queue: Arc<Mutex<Queue>>,
+    sumeragi: Arc<Mutex<Sumeragi>>,
+    blockchain: Arc<Mutex<Blockchain>>,
+    last_round_time: Arc<Mutex<Instant>>,
+    transactions_receiver: Arc<Mutex<TransactionReceiver>>,
 }
 
 impl Iroha {
     pub fn new(config: Configuration) -> Self {
-        let (tx, rx) = mpsc::unbounded();
-        let world = World::new(rx);
-        let torii = Torii::new(
-            &config.torii_url,
-            Sumeragi::new(),
-            Blockchain::new(Kura::new(
-                config.mode,
-                Path::new(&config.kura_block_store_path),
-                tx,
-            )),
-            world,
-        );
-        Iroha { torii }
+        let (transactions_sender, transactions_receiver) = mpsc::unbounded();
+        let (blocks_sender, blocks_receiver) = mpsc::unbounded();
+        let torii = Torii::new(&config.torii_url, transactions_sender, blocks_receiver);
+        let sumeragi = Sumeragi::new();
+        let blockchain = Blockchain::new(Kura::new(
+            config.mode,
+            Path::new(&config.kura_block_store_path),
+            blocks_sender,
+        ));
+        Iroha {
+            queue: Arc::new(Mutex::new(Queue::default())),
+            torii: Arc::new(Mutex::new(torii)),
+            sumeragi: Arc::new(Mutex::new(sumeragi)),
+            blockchain: Arc::new(Mutex::new(blockchain)),
+            transactions_receiver: Arc::new(Mutex::new(transactions_receiver)),
+            last_round_time: Arc::new(Mutex::new(Instant::now())),
+        }
     }
 
-    pub async fn start(&mut self) -> Result<(), String> {
-        self.torii.start().await;
+    pub async fn start(self) -> Result<(), String> {
+        let pool = ThreadPool::new().unwrap();
```

Comment:
I thought we agreed to use `unwrap` only in tests? :smile:

Reply:
already removed)